presto-main/src/test/java/com/facebook/presto/type/TestMapOperators.java

```diff
@@ -335,8 +420,46 @@ public void testJsonToMap()
                 ImmutableMap.of("k1", "{\"7klmnopq\":7,\"6klmnopq\":6,\"5klmnopq\":5,\"4klmnopq\":4,\"3klmnopq\":3,\"2klmnopq\":2,\"1klmnopq\":1}")
         );
 
+        // nested array/map
+        assertFunction("CAST(JSON '{\"1\": [1, 2], \"2\": [3, null], \"3\": [], \"5\": [null, null], \"8\": null}' AS MAP<BIGINT, ARRAY<BIGINT>>)",
+                mapType(BIGINT, new ArrayType(BIGINT)),
+                asMap(
+                        ImmutableList.of(1L, 2L, 3L, 5L, 8L),
+                        asList(asList(1L, 2L), asList(3L, null), asList(), asList(null, null), null)));
+
+        assertFunction("CAST(JSON '{" +
+                        "\"1\": {\"a\": 1, \"b\": 2}, " +
+                        "\"2\": {\"none\": null, \"three\": 3}, " +
+                        "\"3\": {}, " +
+                        "\"5\": {\"h1\": null,\"h2\": null}, " +
+                        "\"8\": null}' " +
+                        "AS MAP<BIGINT, MAP<VARCHAR, BIGINT>>)",
+                mapType(BIGINT, mapType(VARCHAR, BIGINT)),
+                asMap(
+                        ImmutableList.of(1L, 2L, 3L, 5L, 8L),
+                        asList(
+                                ImmutableMap.of("a", 1L, "b", 2L),
+                                asMap(ImmutableList.of("none", "three"), asList(null, 3L)),
+                                ImmutableMap.of(),
+                                asMap(ImmutableList.of("h1", "h2"), asList(null, null)),
+                                null)));
+
+        // invalid cast
         assertInvalidCast("CAST(JSON '{\"true\":\"kittens\"}' AS MAP<BOOLEAN, VARBINARY>)");
         assertInvalidCast("CAST(JSON '{\"[1, 2]\": 1}' AS MAP<ARRAY<BIGINT>, BIGINT>)");
+        assertInvalidCast("CAST(JSON '{\"1\":1, \"01\": 2}' AS MAP<BIGINT, BIGINT>)");
+        assertInvalidCast("CAST(JSON '[{\"1\":1, \"01\": 2}]' AS ARRAY<MAP<BIGINT, BIGINT>>)");
```

Comment:
Make sure all `checkCondition` are covered here. Please assert the error message.